packages/trie/src/db.ts

```diff
@@ -27,9 +34,53 @@ export class DB {
    * @param leveldb - An abstract-leveldown compliant store
    */
   constructor(leveldb?: LevelUp) {
+    // Roots of trie at the moment of checkpoint
+    this.checkpoints = []
+
     this._leveldb = leveldb || level()
   }
 
+  /**
+   * Is the DB during a checkpoint phase?
+   */
+  get isCheckpoint() {
+    return this.checkpoints.length > 0
+  }
+
+  /**
+   * Adds a new checkpoint to the stack
+   * @param root
+   */
+  checkpoint(root: Buffer) {
+    this.checkpoints.push({ root, revertOps: [] })
+  }
+
+  /**
+   * Commits the latest checkpoint
+   */
+  commit() {
+    const { root, revertOps } = this.checkpoints.pop()!
+    // On nested checkpoints put the revertOps on the parent
+    // stack in case there is a revert
+    if (this.isCheckpoint) {
+      this.checkpoints[this.checkpoints.length - 1].revertOps.concat(revertOps)
+    }
+    return root
+  }
+
+  /**
+   * Reverts the latest checkpoint
+   */
+  async revert() {
+    const { root, revertOps } = this.checkpoints.pop()!
+    await this.batch(revertOps.reverse())
```

Comment:
This seems to be reversed because of the logic in the `put` method; we first want to `del` the key and then `put` the old value in the corresponding key - right?

Reply:
Yeah, correct, this is just a design decision, the alternative would have been to always prepend to the array which didn't seem right to me.